applications/FluidDynamicsApplication/python_scripts/cfl_output_process.py

```diff
@@ -1,11 +1,19 @@
 # Importing the Kratos Library
 import KratosMultiphysics
 import KratosMultiphysics.FluidDynamicsApplication as KratosCFD
+from KratosMultiphysics.kratos_utilities import CheckIfApplicationsAvailable
 
 # other imports
 from KratosMultiphysics.time_based_ascii_file_writer_utility import TimeBasedAsciiFileWriterUtility
 
-from statistics import mean, stdev
+if CheckIfApplicationsAvailable("StatisticsApplication"):
+    from KratosMultiphysics.StatisticsApplication import SpatialMethods as spatial_methods
+else:
+    msg = "CFLOutputProcess requires StatisticsApplication which is not found."
+    msg += " Please install/compile it and try again."
+    raise Exception(msg)
```

Comment:
I wouldn't introduce such dependency. As far as I know the process can be used without this new feature.

Reply:
The problem is, existing implementation was on python level, which is really expensive for large MPI problems. That was the reason for Statistics application.

Reply:
StatisticsApplication can be thought of as a utility library.

Reply:
But right now you are making mandatory to have the  `StatisticsApplication` to use something that before could be used without it. I'm not completely convinced about this...

I think it can be thought of as an optional utility library. Hence, I'd throw an error if the user wants to use the statistics without having the `StatisticsApplication`. One may want to just print the CFL without any statistics as it was right now.

Reply:
As I see, this existing implementation always calculates statistics of CFL, one cannot use it just to calculate CFL, that's why I removed redandant codes. I understand that this makes dependency, in that case I can make this process to have existing implementation and only run with C++ Backend If StatisticsApplication ist available.?

Reply:
Having two implementations is a no go for me. If there is no way to keep some functionalities without introducing the complete dependency on the `StatisticsApplication` fine. In any case, I think that the max, which is one of the more useful variables for the user, and min values can be efficiently computed without introducing such dependency.

Reply:
Hi,

The idea to have only python math and/or statistic library initially was to be independent of other (although more performant) libraries (such as numpy). Now the StatisticsApplication would internally solve this efficiency issue. To give you some numbers (very recent simulation):

Fraction Step CFD (no predictor corrector) with 7.5 mio elements
On 192 cores MPI so circa 40k elements/cores

**Without CFL output at each time step:**
 15 min - 122 steps (circa 2  steps per each time step for fluid velocity solver convergence)
==>> 7.4s/step

**With CFL output at each time step:**

_Old way - with python syntax:_
15 min - 12 steps (circa 2 steps per each time step for fluid velocity solver convergence)
-> 75s/step
-> from which 75 - 7.4= 67.6s/CFL solve

_New way - with c++ syntax:_
15 min - 110 steps (circa 2 steps per each time step for fluid velocity solver convergence)
->8.2s/step
-> from which 8.2-7.4 = 0.8s/CFL solve

So the CFL Output Process time goes from circa **70s to 0.7s (factor 100 decrease)** for each solve... Mind you, the statistics is not very heavy, the problem is with the python list storing CFL values from each element. Starting from around 3-5mio element CFD simulations (which is definitely not uncommon), the performance is really bad using pure python (perhaps even numpy arrays would not be that much better).

If you look at the numbers above, in the old (current way) the Fractional Step fluid solve (with 2 iterations - for 7.5 mio elements!!!) is 10 times faster per step than the CFL output evaluation. This is also what we observed with another test case with @AndreasWinterstein .

In general, I would suggest introducing the usage of the StatisticsApplication wherever possible because:

- it is very efficient
- it is lightweight
- it is an internal dependency (for example FSI depende internally on MeshMoving)
- it a matter of "simply" having it compiled
- it could potentially constitute a type of utility (perhaps getting into core functionalitites? Here @philbucher @KratosMultiphysics/technical-committee could drop a note)

I do not see any disadvantages. The current pure python implementation came due to a restricited necessity.

Reply:
Just to clarify that I'm not opposing to the introduction/use of the `StatisticsApplication`. My point is that we need to be really careful in how we introduce it (we don't want another mess that lasts for years and takes even more years to be amended as it used to happen in the `TrilinosApplication` not long ago).

Having said this, and considering the numbers you post above (which are indeed really impressive), I would try to answer some of your points one-by-one:

>     * it is an internal dependency (for example FSI depende internally on MeshMoving)
> 
I partially agree with this. I have to comments in here. The first is that you can't do ALE based FSI without the `MeshMoving` (nor the FluidDynamics, Structural). Hence this dependency comes with the physics you want to solve in this app. Conversely, you can perfectly do CFD without the statistics, and actually this is what the vast majority of users of the `FluidDynamicsApplication` do.
Besides this, the `FluidDynamicsApplication` is understood as a core application. Therefore we need to discuss and be really careful with the dependencies we introduce.

I understand that right now is just a simple Python dependency, but as this is the first use case of the `StatisticsApplication` I want to open the discussion to find the most long-term efficient way to introduce this new app.
At this point I think it is very important to clarify if you foresee any c++ future dependency.

>     * it a matter of "simply" having it compiled
> 
This is the point. Right now it's a matter of having it compiled. But would it be always in this way? (I recall to my c++ comment above).

>     * it could potentially constitute a type of utility (perhaps getting into core functionalitites? Here @philbucher @KratosMultiphysics/technical-committee could drop a note)
> 
The core is way large right now. Taking this into account as well as the purpose-specific conception of the app, I'd rather keep it as a separate application.

I briefly discussed with @sunethwarna the possibility of finding a way to introduce the `StatisticsApplication` in a more elaborated way to avoid future issues.

Reply:
Hi @rubenzorrilla I opened the #6765 issue regarding this. As you mentioned, we might end up with some compiler level dependencies, and I understand hacing such dependencies gonna be hard to maintain, and not the most elegant way. If we can come to a proper reciepe as you said, which will not have redundant codes as well as smart way of doing application dependencies that would be perfect.
Kratos is made to be modular in the sense that several applications can be used to gether ryt?, So in that case we need to have proper dependency handling method. (I am foreseeing this in RANSApplication, which will be in compiler level, and FluidDynamicsApplication with "integration_point_statistics_process.cpp")

Reply:
For the specific purpose:
CFL Output Process can for now only be used from the Python level and not from GUI, so perhaps it can be expected of the advanced user to also have StatisticsApplication compiled

In general: the discussion of where to place StatisticsApplication is out of my reach, whether it should be an application, utility or part of core is, as you said, part of a broader discussion.

Some advantages (and necessities) which I see:

    advanced and industry users will need and expect fast statistics (for convergence monitors, as dependencies/sub-functionalities of other processes, etc.)
    the more of these processes are used, the more relevant the computational efficiency (or lack of it) will be
    I am mostly talking from the perspective of CFD, FSI and time-dependent CSD, but probably a lot of applications could benefit from it

Reply:
without going into detail, IMO none of the "main"/core apps can have (C++) dependencies on other apps (except Trilinos but this is the basis for all our MPI-capabilities right now)
Also "basic" features of those apps cannot depend unconditionally on other apps (ofc there are exceptions such as MeshMoving and Mapping for FSI, or linear solvers being in a different app)

For more advanced use cases (such as this one which has huge advantages) to me it makes perfect sense though to add the (python) dependency for things like this, where the scope of features is very clearly separated (FluidApp does fluid, StatisticsApp does statistics)
This way not only different apps can use the features, but also we don't overload the big apps even more!

Regarding the interface IMO this cannot be a blocker for the development of new features and functionalities. I am pretty sure there will be a way to make this available if needed